### ISSUE TITLE
research: MultiStartProdigy JAX compile times — measure (MGE + meshes), make fast

### ISSUE THREAD

**Issue by Jammy2211** [PyAutoLabs/autolens_profiling#93]
## Overview

MultiStartProdigy is now the endorsed search for MGE and pixelized (rectangular kernel-CDF / knn / delaunay) source fits, but its JAX jit compile cost has never been measured as such: the closed compile-time arc (#71→#77) certified cache + autotune-off for *single-start* transforms only, and the mesh model types were never compile-probed at all. A later follow-up (multi-band section of `scripts/misc/jax_compile/README.md`) found the MultiStartProdigy production transform — an in-XLA `lax.map(value_and_grad, batch_size)` scan — is compile-intractable (OOM-killed) on a 1-core CPU host, while a Python loop over `vmap` chunks ("pyloop") compiles at single-fit cost. This task measures the prodigy transform × model-type matrix and, if confirmed, ships the pyloop lever in PyAutoFit so all four cells compile fast.

## Plan

- **Phase A — measure (autolens_profiling):** cold + warm compile of {mge, pixelization, knn, delaunay(_matern)} × {vag, vmap_vag n=16, laxmap_vag bs=4, pyloop_vag bs=4} via `scripts/misc/jax_compile/probe.py`; local idle CPU first, A100 rows via `hpc/` where CPU numbers indicate. Results committed as JSON rows + README findings/verdict.
- **Phase B — fix (PyAutoFit, gated on Phase A):** hoist MultiStartProdigy/Adam batching out of XLA — Python loop over one jitted `vmap(batch_size)` chunk; `batch_size=None` full-vmap path unchanged; numerically identical results. Before/after compile numbers per cell recorded in autolens_profiling.
- Library PR first; profiling-results PR behind it. Settings-first discipline: confirm cache + autotune-off are doing their part per cell before touching source.

<details>
<summary>Detailed implementation plan</summary>

### Affected Repositories
- autolens_profiling (primary — measurement + results)
- PyAutoFit (Phase B lever, gated on measurements)

### Branch Survey
| Repository | Current Branch | Dirty? |
|-----------|---------------|--------|
| ./autolens_profiling | main | 1 untracked (auto-simulated `dataset/imaging/jwst_lw/`, harmless) |
| ./PyAutoFit | main | clean |

**Suggested branch:** `feature/multistart-prodigy-compile`
**Worktree root:** `~/Code/PyAutoLabs-wt/multistart-prodigy-compile/`
**Work Classification:** Both (workspace measurement + library fix)

Coordination flag: parallel task `preserve-in-zip-replace-member` (PyAutoFit#1414) has a PyAutoFit feature branch in its own worktree; no formal `active.md` repo claim yet, but the Phase B PyAutoFit PR should coordinate/serialize with it.

### Implementation Steps (Phase A)
1. Verify `scripts/misc/jax_compile/probe.py --model-type` accepts the pix mesh types added in PR#91 (`pixelization`, `knn`, `delaunay`, `delaunay_matern` in `scripts/misc/searches/_setup.py`); wire any missing plumbing (adapt images / AdaptSplit-family regularization already handled by `_setup.py`).
2. Run the matrix on an idle local CPU (compile is host-CPU-bound; idle-machine discipline per the README): each cell cold (fresh `--cache-dir`) then warm (same dir, fresh process), transforms `vag,vmap_vag,laxmap_vag,pyloop_vag`, n_starts 16 / batch_size 4 (production knobs from the prodigy campaign).
3. Delaunay caveat: the qhull host callback may not tolerate AOT compile of the batched transform — record `delaunay_matern` as the family representative if plain `delaunay` blocks, and note it.
4. Append JSON rows under `scripts/misc/jax_compile/results/<hardware>/<model_type>.json`; add a findings section ("MultiStartProdigy transform census") + verdict to the README.
5. A100 rows via `hpc/` submit path for cells where CPU numbers say GPU matters (rect jvp disproportion is an open follow-up item from wsdev#117).

### Implementation Steps (Phase B — only if Phase A confirms)
6. `PyAutoFit/autofit/non_linear/search/mle/multi_start_gradient/search.py`: replace the `jax.lax.map(_value_and_grad, params, batch_size=)` closure (~line 259) with a Python loop over one jitted `vmap` chunk of `batch_size` starts, concatenating results; `batch_size=None` path (single full vmap) unchanged. `batch_size` semantics and results identical by construction.
7. Unit tests numpy-only (no JAX in library unit tests); numerical parity + compile before/after validated via the profiling probe and the existing prodigy cells (`scripts/imaging/searches/multi_start_prodigy/*.py`).
8. Re-run the affected probe cells; record before/after in the README section.

### Key Files
- `autolens_profiling/scripts/misc/jax_compile/probe.py` — the AOT-split compile probe (transform axis already includes `laxmap_vag` / `pyloop_vag`)
- `autolens_profiling/scripts/misc/jax_compile/README.md` — findings home; multi-band section documents the scan-compile blow-up and names the pyloop lever
- `autolens_profiling/scripts/misc/searches/_setup.py` — model/dataset builders incl. `knn` / `delaunay_matern` (PR#91)
- `PyAutoFit/autofit/non_linear/search/mle/multi_start_gradient/search.py` — `MultiStartOptimizer._fit` batching closure (Phase B target)

### Recorded overrides
- Feature Agent scored the prompt too-large (score 12, repo-count heuristic) and proposed a generic 4-phase split; overridden to the natural 2-phase measure→fix shape (measurement is one instrument run over existing cells; the fix is one bounded closure change).
</details>

## Original Prompt
<details>
<summary>Click to expand starting prompt</summary>

# Profile MultiStartProdigy JAX compile times (MGE + pix meshes) and make them all compile fast

Type: research
Target: autolens_profiling
Repos:
- @autolens_profiling
- @PyAutoFit
Difficulty: medium
Autonomy: supervised
Priority: high
Status: formalised

## Original request (verbatim)

> We have concluded that MultiStartProdigy is good to use on both sources with
> an MGE and rectangular / knn / delaunay mesh (see recent notes in
> autolens_workspace and autolens_workspace_developer and autolens_profiling).
> However, I think MultiStartProdigy can still have long jax jit compile times,
> and if it doesnt we have shown this for meshes. Can you do profiling of this
> (building on any previous work) and aim to make them all compile fast. Use
> the search pyautobrain conductor, make sure results are stored in
> autolens_profiling.

## Prior work to build on (do not re-derive)

- The compile-time arc #71→#74→#77 (autolens_profiling `scripts/misc/jax_compile/README.md`) is CLOSED for single-start transforms: persistent cache (PyAutoNerves#128) + autotune-off (#132) give worst cold ~35 s / warm 5–9 s on the A100 census cells. Verdict stands: settings first, no likelihood restructuring.
- The multi-band follow-up (same README, final section) measured the **MultiStartProdigy production transform** — `jax.lax.map(value_and_grad, batch_size=)` (PyAutoFit `multi_start_gradient/search.py`) — and found the in-XLA scan is the compile killer: cold-compile intractable / OOM-killed on a 1-core CPU host, while `pyloop_vag` (Python loop over `vmap` chunks, same vmap width) compiles at single-fit cost (166 s). Named "the leading source lever"; validated on the homogeneous 4-band MGE graph only.
- `probe.py` already has the transform axis (`vag`, `vmap_vag`, `laxmap_vag`, `pyloop_vag`) and `scripts/misc/searches/_setup.py` has all four model types: `mge`, `pixelization` (rect kernel-CDF), `knn`, `delaunay` / `delaunay_matern` (added in autolens_profiling PR#91).
- Prodigy campaign context (wsdev#117 / profiling PR#91): production knobs are 16 starts / batch_size 4; rect throughput anomaly (~17× knn per step vs 4.5× forward-eval — jvp disproportion) is an open A100 follow-up item.
- The stale broad draft `draft/feature/autolens_profiling/jax_compile_time_profiling.md` predates the arc close (its autotune section was overturned); this prompt supersedes its measurement scope for the search transform. Fold or retire it when this ships.

## Scope

1. **Measure** (autolens_profiling, results committed under the repo's dashboard conventions): cold + warm compile of the MultiStartProdigy transform matrix — {mge, pixelization, knn, delaunay(_matern)} × {vag, vmap_vag(n=16), laxmap_vag(bs=4), pyloop_vag(bs=4)} — local CPU first; A100 rows via the hpc/ submit path where the CPU numbers say they matter.
2. **Fix** (PyAutoFit, only if the measurements confirm the lever): make MultiStartProdigy/MultiStartAdam batching compile fast — leading candidate is hoisting the batch loop out of XLA (`pyloop` pattern: Python loop over `vmap(batch_size)` chunks), keeping `batch_size` semantics and results identical. Settings-first discipline applies: confirm cache + autotune-off are already doing their part per cell before touching source.
3. **Record**: results + verdict live in autolens_profiling (jax_compile results or searches results tier as fits the repo layout); library change ships through the normal PyAutoFit PR flow.

</details>

**Comment by Jammy2211** [comment]
**Phase A underway.** Worktree `multistart-prodigy-compile` created (autolens_profiling @ `feature/multistart-prodigy-compile`); wiring smoke passed (knn/jit: trace 7.9 s, compile 9.2 s, local CPU). Full census sweep launched sequentially on an idle 1-core host: {mge, pixelization, knn, delaunay_matern} × {jit, vag, vmap_vag, pyloop_vag, laxmap_vag} × cold/warm, n_starts 16 / batch_size 4, per-run timeout caps (a laxmap DNF is recorded as a finding). Results append under `scripts/misc/jax_compile/results/local_cpu/`. Phase B (PyAutoFit pyloop batching in `MultiStartOptimizer._fit`, covers Prodigy+Adam) is designed and gated on these numbers; it also serialises behind PyAutoFit#1414's merge.

**Comment by Jammy2211** [comment]
## Hand-off / resume state (2026-07-28, end of session)

**Phase A is COMPLETE and shipped as PR #94** (results-only, `pending-release` labelled).

### The answer

**Single-band MultiStartProdigy compile time is a non-problem** on every endorsed model type — nothing needed making fast. Measured {mge, pixelization (rect kernel-CDF), knn, delaunay_matern} x {jit, vag, vmap_vag, pyloop_vag, laxmap_vag}, cold/warm, at production knobs (16 starts, batch_size 4):

| tier | worst cold (trace+compile) | warm |
|---|---|---|
| 1-core 15 GB laptop (worst case) | ~3.5 min (MGE vmap_vag; meshes 35-75 s) | 2-4 s |
| RAL 32-core / 128 GB node | ~75 s (MGE laxmap; meshes 26-33 s) | 1-2 s |

The multi-band `lax.map` compile explosion (intractable / compile-OOM, previous README section) **does not reproduce single-band** — a single-band likelihood body is benign as a scan body. So the planned **phase B PyAutoFit pyloop change is an evidence-based NO-GO**; that lever stays reserved for the multi-band `FactorGraphModel` case. No PyAutoFit branch was created, so the PyAutoFit#1414 serialisation concern is moot.

### What the census found instead

1. **The delaunay family never hits the persistent compilation cache** — warm compile identical to cold (16-33 s in *every* process), reproduced to the decimal on two independent hosts, while knn/rect/mge warm to ~1 s in the same job. knn (pure-JAX, no host callback) is the control; prime suspect is the qhull `pure_callback` embedding a process-specific descriptor in the HLO. **This is the one real compile defect** — follow-up prompt filed at `PyAutoMind draft/research/autoarray/delaunay_callback_persistent_cache_miss.md`.
2. **Rect batched gradients are memory-bound, not compile-bound**: ~9.2 GB per start in the jvp -> width 4 = 37 GB (OOM on a 15 GB laptop; RAL job MaxRSS confirmed 39.4 GB), width 16 = 147 GB (fits nowhere). **`batch_size=4` is load-bearing, not a tuning nicety.**
3. Rect's real cost axis is **throughput** (~310 s per 16-start step on 32 CPU cores, matching the campaign's ~5.7 min/step) — already on the A100 follow-up list from wsdev#117.

### Resume checklist (in order)

1. **A100 rows** — RAL job **331380** is queued behind an external multi-day array (gpu-2 saturated, gpu-1 down). When it runs: rsync `/mnt/ral/jnightin/autolens_profiling_census/scripts/misc/jax_compile/results/`, filter tag `prodigy-census-a100-*`, append to the README table + results JSON, commit to the same branch. **Confirmatory only — do not block the merge on it** (#77 already put single-band A100 compiles at seconds-to-30 s).
2. **Merge PR #94 and close this issue** (human decision).
3. **Route the delaunay cache-miss follow-up** via `/start_dev draft/research/autoarray/delaunay_callback_persistent_cache_miss.md` — the actionable outcome of this census.
4. **Post-merge cleanup** — worktree `~/Code/PyAutoLabs-wt/multistart-prodigy-compile`, branch, and the `complete/2026/07` record.

Full detail, tables and provenance live in `scripts/misc/jax_compile/README.md` ("MultiStartProdigy transform census" section) — a fresh session can start from that file plus the `active.md` entry with no prior chat context.

### Reproduce
```bash
# one cell, from the autolens_profiling root
python3 scripts/misc/jax_compile/probe.py --model-type knn --transforms laxmap_vag \
  --n-batch 16 --batch-size 4 --cache-dir /tmp/jcc_knn --tag repro
```
Run cold (fresh `--cache-dir`) then warm (same dir, fresh process). Compile is host-CPU-bound, so use an idle machine or a dedicated allocation — loaded-host numbers were historically wrong by up to 7x. RAL submit scripts: `/mnt/ral/jnightin/pixgrad_logs/census_{cpu,gpu}.sbatch`.

**Comment by Jammy2211** [comment]
**Merged as [PR #94](https://github.com/PyAutoLabs/autolens_profiling/pull/94) (`1088f41`).** Closing.

**Verdict: `af.MultiStartProdigy` compile time is a non-problem on every endorsed model type** — MGE, rectangular kernel-CDF, KNN and Delaunay. Nothing needed speeding up: ≤75 s cold and ≤2 s warm on a 32-core node; ~3.5 min worst case on a deliberately worst-case 1-core laptop. The multi-band `lax.map` compile explosion does **not** reproduce single-band, so the planned PyAutoFit pyloop change was a no-go on the evidence and no library branch was ever created.

What the census produced instead:
1. **Delaunay never hits the persistent compilation cache** — full recompile every process, confirmed on two hosts (KNN is the pure-JAX control). Routed as a follow-up.
2. **Rectangular `batch_size=4` is load-bearing for memory** (~9.2 GB per start; width 16 = 147 GB, fits nowhere) — quantitative basis for the campaign's config.
3. **A repaired data-integrity bug**: `results/local_gpu_*/mge.json` had a stray trailing `]` since 270e5cc that would have crashed the next A100 mge probe run.

**Documented at:** [`results/notes/multistart_prodigy_compile_census.md`](https://github.com/PyAutoLabs/autolens_profiling/blob/main/results/notes/multistart_prodigy_compile_census.md), linked from the main README, which now also carries `jax_compile/` in its section index (previously unreachable from the front door) and current gradient/compile-time standing conclusions in place of the stale "JAX gradients — out of scope" section.

**Left open deliberately:** the A100 tier. Job 331380 ran while the node's A100s were externally saturated; `cuInit` returned `CUDA_ERROR_NO_DEVICE` and JAX silently fell back to CPU, so those rows were discarded rather than committed (trap documented in both READMEs). A100 rows are confirmatory only — #77 already put single-band A100 compiles at seconds-to-30 s and both CPU tiers agree the verdict isn't tier-sensitive. Re-run with `sbatch /mnt/ral/jnightin/pixgrad_logs/census_gpu.sbatch` when a GPU node is free.
